frontend/src/main/scala/bloop/exec/Forker.scala

```diff
@@ -188,43 +190,46 @@ object Forker {
   }
 
   /**
-   * Return an array of lines from a process buffer and a no lines buffer
-   *
-   * The no lines buffer keeps track of previous messages that did not contain
-   * a new line, it is therefore mutated. The buffer is the logs that we just
-   * received from our process.
+   * Performs specified operation on each line from the process [[input]].
+   * Segment without a new line is carried over using the [[carry]] buffer.
    *
-   * This method returns an array of new lines when the messages contain new
-   * lines at the end. If there are several new lines in a message but the last
-   * one doesn't, then we add the remaining to the string builder.
-   *
-   * @param buffer The buffer that we receive from NuProcess
-   * @param remaining The string builder bookkeeping remaining messages without new lines
-   * @return An array of new lines. It can be empty.
+   * @param input containing lines on which the [[op]] must be performed
+   * @param carry The string builder bookkeeping remaining message without new lines
    */
-  private[bloop] def linesFrom(buffer: ByteBuffer, remaining: StringBuilder): Array[String] = {
-    val bytes = new Array[Byte](buffer.remaining())
-    buffer.get(bytes)
-    val msg = new String(bytes, StandardCharsets.UTF_8)
-    // TODO what when we attach to a process on different system?
-    val newLines = msg.split(System.lineSeparator, Integer.MAX_VALUE)
-    newLines match {
-      case Array() => remaining.++=(msg); Array.empty[String]
-      case msgs =>
-        val msgAtTheEnd = newLines.apply(newLines.length - 1)
-        val shouldBuffer = !msgAtTheEnd.isEmpty
-        if (shouldBuffer)
-          remaining.++=(msgAtTheEnd)
-
-        if (msgs.length > 1) {
-          if (shouldBuffer) newLines.init
-          else {
-            val firstLine = newLines.apply(0)
-            newLines(0) = remaining.mkString ++ firstLine
-            remaining.clear()
-            newLines.init
+  private[bloop] def onEachLine(input: ByteBuffer, carry: StringBuilder)(
+      op: String => Unit
+  ): Unit = {
+    val bytes = new Array[Byte](input.remaining())
+    input.get(bytes)
+    val newMessage = new String(bytes, StandardCharsets.UTF_8)
+
+    if (newMessage.nonEmpty) {
+      val msg = new StringBuilder()
+        .append(carry)
+        .append(newMessage)
+      carry.clear()
+
+      @tailrec
+      def traverseLines(start: Int): Unit = {
+        if (start < msg.length) {
+          val lineEnd = msg.indexOf(System.lineSeparator(), start)
+          if (lineEnd < 0) {
+            // JVM send the JDI notification with "\n" as newline delimiter even on windows
+            if (CrossPlatform.isWindows && msg.startsWith(DebugSessionLogger.JDINotificationPrefix)) {
+              op(msg.stripLineEnd)
```

Comment:
The message could be partially filled in the buffer, so I'm not sure this line would work as expected. I think we could make it work by doing `val jdiIdx = msg.indexOf(DebugSessionLogger.JDINotificationPrefix)` and then checking that LF exists with `msg.indexOf("\n", jdiIdx)`. Then, I think `op(msg.stripLineEnd)` should be correct